proc.c

```diff
@@ -3003,9 +3003,12 @@ method_inspect(VALUE method)
     else {
         mklass = data->klass;
         if (FL_TEST(mklass, FL_SINGLETON)) {
-            do {
-               mklass = RCLASS_SUPER(mklass);
-            } while (RB_TYPE_P(mklass, T_ICLASS));
+            VALUE v = rb_ivar_get(mklass, attached);
+            if (!(RB_TYPE_P(v, T_CLASS) || RB_TYPE_P(v, T_MODULE))) {
+                do {
+                   mklass = RCLASS_SUPER(mklass);
+                } while (RB_TYPE_P(mklass, T_ICLASS));
+            }
```

Comment:
Would using a `while {}` loop instead of `do {} while` also solve it, or is the attached check necessary?

Can singleton classes sometimes be `T_ICLASS` too? I thought T_ICLASS is only for included modules.

Reply:
I don't believe `mkclass` can be a `T_ICLASS` before the loop, so switching to a `while` loop would turn this into a no-op.  To be fair, I didn't actually try that, so my understanding may be incorrect.